### ISSUE TITLE
Proof by pointing on tablets?

### ISSUE THREAD

**Issue by herbelin** [coq/coq#6056]
Hi, I feel that [proof-by-pointing](https://www-sop.inria.fr/croap/ctcoq/help/pbp.html) (an old concept from CtCoq/[pcoq](http://www-sop.inria.fr/lemme/pcoq/)) would be pretty useful on tablets. I wonder how we could start some experiences about that, on one of our favorite user interfaces.

For instance, without even thinking as moving things with the mouse or the finger yet, what would be missing in the current protocol and/or GUIs in order that a click, or a few-second-non-sliding-finger on an hypothesis, or on a subterm opens a menu with a list of possible actions related to this hypothesis or subterm?

Similary, can we rethink with a 20 years hindsight at a proof-by-pointing language interpretable as a tactic script in the current sense, even if only for a few actions/gestures to start with?

**Comment by psteckler** [comment]
FWIW, I seem to recall running some OCaml code on Android a couple of years ago. I think I used this OCaml cross-compiler: https://github.com/ocaml-cross/opam-cross-android

**Comment by herbelin** [comment]
Actually, I was motivated by this kind of experience that you are mentioning, and in particular by a recent discussion on coq-club which maybe you missed as it was during your vacations ([here](https://sympa.inria.fr/sympa/arc/coq-club/2017-10/msg00050.html)).

**Comment by psteckler** [comment]
The discussion there seems limited to using Coq with a terminal emulator on Android, while you'd need some graphical interface to do proof-by-pointing. An Android-native interface would be best (I've written one, not too hard).

**Comment by herbelin** [comment]
> The discussion there seems limited to using Coq with a terminal emulator on Android, while you'd need some graphical interface to do proof-by-pointing.

Ah, I may have missed that point. Indeed, it seemed to have been using `emacs -nw`.

>  An Android-native interface would be best (I've written one, not too hard).

I'm curious, can you give more details?

**Comment by psteckler** [comment]
> Indeed, it seemed to have been using emacs -nw

Yes, using ssh to get to a desktop machine, if I recall that message.

> I'm curious, can you give more details?

Android has XML files that specify the widgets and layout you want to appear in your interface. There are graphical tools to generate the XML files; sometimes, you have to hand-massage them to get them just right.

Then you write some handlers (in Java or another JVM language) to interact with the widgets.

Coq could be started as a process from Java (or other language), and its I/O sent to the program.

I found the hardest thing is trying to deal with the different sizes of device screens.

**Comment by herbelin** [comment]
Hum, it is hard to me to visualize what you obtained. Do you mean that you could e.g. get a widget where you can send text to a coqtop process and the output be shown in an output buffer or something like that?

BTW, what about JsCoq? I couldn't get it on my iPhone (because of unsupported Safari IIRC), but does it work on Android?

**Comment by psteckler** [comment]
Yes, something like that, and you can have menus, scrollbars, text entry fields, etc.

I have not tried JsCoq on Android, although I'd think it would work in any browser with JavaScript. Android uses Chrome as its default browser.

**Comment by herbelin** [comment]
But you would need a structured output from Coq to be able to do some proof-by-clicking...

**Comment by psteckler** [comment]
Or a parser.

**Comment by herbelin** [comment]
But a parser is difficult: extensible grammar, dependence on databases for scopes, implicit arguments, disambiguation of variables/constants, ...

**Comment by psteckler** [comment]
OK, having that knowledge on the Coq side would be better, indeed.

**Comment by ejgallego** [comment]
As you know I have been working on this for a while, and indeed using HTML is the best approach IMO (the backend can be JsCoq or regular Coq, the display HTML).

However I'd like to do it properly, and IMHO to do that properly a lot of work remains to be done. However we have advanced a lot, in particular the `CAst.t` thing makes much easier to embed metadata in the AST, but some of more tricky parts do remain, see the corresponding CEP. In particular the friction between `constrexpr` and `glob_term` is still too large IMHO to get some of the most sensible stuff done.

Also, we'd need a proper target document type. we are not so far from it, but I think more work is needed, see a bit the discussion on #505 . And at last, I believe a new term/goal printer to such box-aware format is needed.

**Comment by psteckler** [comment]
Android has an HTML widget that supports Javascript, so you could have a dedicated Coq app that uses it for display (and add whatever else you'd want, like buttons, menus).

**Comment by ejgallego** [comment]
It should run in any browser, even in modern Safari. I'm sorry I broke it in Safari but manpower is very limited, there is no fundamental reason it should not work well in Safari.

**Comment by herbelin** [comment]
@ejgallego (penultimate comment): thanks, that is exactly the update of information I was expecting.
